File: problem23.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
import heapq
class Item:
    def __init__(self, node):
        self.val=node.val
        self.next=node.next
        
    def __lt__(self, other):
        return self.val<other.val
        
class Solution:
    def mergeKLists(self, lists: List[ListNode]) -> ListNode:
        heap = []
        head=ListNode()
        curr=head
        #insert first element from each list and push into heap
        for i in range(len(lists)):
            if lists[i]:
                l=Item(lists[i])
                heappush(heap,l)
        #continue till heap has elements. Pop the min element and push the next element of current minimum element.
        while(heap):
            minelement = heapq.heappop(heap)
            curr.next = ListNode(minelement.val)
            curr=curr.next
            if minelement.next:
                l = Item(minelement.next)
                heapq.heappush(heap, l)
        #return the head once the heap is empty.
        return head.next
    
#time: O(NklogK)
#space: O(K)

File: problem23.cs
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     public int val;
 *     public ListNode next;
 *     public ListNode(int val=0, ListNode next=null) {
 *         this.val = val;
 *         this.next = next;
 *     }
 * }
 */
public class Solution {
    public ListNode MergeKLists(ListNode[] lists) {
        
        if(lists.Length==0)
            return null;
        
        if(lists.Length==1)
            return lists[0];
        
        ListNode head = new ListNode();
        ListNode current = head;
        PriorityQueue<ListNode, int> minHeap = new PriorityQueue<ListNode, int>();
        
        foreach(ListNode h in lists)
        {
            if(h!=null)
                minHeap.Enqueue(h, h.val);
        }
        
        while(minHeap.Count>0)
        {
            ListNode minElement = minHeap.Dequeue();
            if(minElement.next!=null)
                minHeap.Enqueue(minElement.next, minElement.next.val);
            current.next = minElement;
            current=current.next;
            current.next=null;
        }
        return head.next;
    }
}

//time: O(NklogN)
//space: O(N)
